Shared storage

TheW00Xer opened the pull request:
Implementation of shared storage system. App will be using pictures that are already saved in device's memory.

App displays pictures from device memory but on tested device the displayed image was sometimes blank black screen.
When trying to display video App shows "Can't play this video" notification.

App doesn't ask for Storage permission automatically so in order to work, user need to give permission manually through android settings.
Later there will be implemented a pop-up window? / notification, that will prompt user to give App storage permission.

r-oung commented:
> App displays pictures from device memory but on tested device the displayed image was sometimes blank black screen.
> When trying to display video App shows "Can't play this video" notification.
What was the URI that it was trying to open?

When there's an error, look at `logcat` and use `Log.i()` and/or breakpoints to debug your code. If you don't understand the error, try searching [stackoverflow](https://stackoverflow.com/), and if you still don't find what you're looking for, paste the error here.

Comments:
- Keep in mind that your `targetSdk` is 32 (see app `build.gradle`), which is [Android 12](https://en.wikipedia.org/wiki/Android_version_history). This is okay for this app. I just want to let you know that it may not necessarily work on temi, which uses Android 6.
- There is some text in your `MainActivity.java`, which should be in `res/` folder.
- Make sure you thoroughly test edge cases, e.g.
  - What happens if there are folders (not just files) in the directory?
  - What happens if it's an empty directory?
  - What if the `Pictures` and `Movies` directory doesn't exist?

TheW00Xer commented:
> 
Thank you for your feedback!

I've updated the code. Activity displaying pictures works well now. The issue was happening becuse it was trying to read folders as files or other files that are not exactly graphic files so I've implemented a File name filter that incudes file extensions that are usable and displayable by the app. 

As for Display video, I've tried to implement same filter but failed. I'll investigate it further.

I think I've automatically upgraded the graddle to such level. Should i change it and make sure it's compatible with Temi as well? So far I've been testing it only on my phone which is running Android 9.

r-oung commented:
> I think I've automatically upgraded the graddle to such level. Should i change it and make sure it's compatible with Temi as well? So far I've been testing it only on my phone which is running Android 9.

This is just an exercise. So as long as you are aware of it, you can leave it. In the future, if you try running your app on another device, you'll know where to look.

I'll review it again after you get the video working and when you think it's "finished".

TheW00Xer commented:
@r-oung Can you please review the code and give me some tips/solutions for my issues please? 
Namely these two:
1. the fact that when the app looks for files it tries to access sub-folders. I would like to Exclude the sub-folders or include them but enable the app to actually read pictures from within instead of displaying black screen.
2. how to properly use an array(list) of multiple "usable" file extension as filter. Since my filter does not work properly when I try to use the array.

r-oung commented:
>     1. the fact that when the app looks for files it tries to access sub-folders. I would like to Exclude the sub-folders or include them but enable the app to actually read pictures from within instead of displaying black screen.
If you want to read images from sub-folders, and their sub-folders, etc. it's called a "recursive" search. Google:
```
java list files recursively
```

There are various ways to do this, but one way is to use a method called `Files.walk`.

>     2. how to properly use an array(list) of multiple "usable" file extension as filter. Since my filter does not work properly when I try to use the array.
I don't know of a Java method that takes an array and filters by extension. In your code, you use `endsWidth`, but that only takes a string [ref](https://docs.oracle.com/javase/8/docs/api/java/lang/String.html#endsWith-java.lang.String-) not an array.

r-oung reviewed (commented):
Sorry, it seems I never submitted these comments.

TheW00Xer commented:
> Sorry, it seems I never submitted these comments.

No problem.

r-oung commented on app/src/main/java/com/example/feelingluckytoday/DisplayPictureActivity.java:
```diff
@@ -1,33 +1,81 @@
 
 public class DisplayPictureActivity extends AppCompatActivity {
 
+    @RequiresApi(api = Build.VERSION_CODES.O)
```
Do you need this? What is this for?

TheW00Xer replied:
This was forgotten fragment of code left over after I was trying something. 
Erased.

r-oung commented on app/src/main/java/com/example/feelingluckytoday/DisplayPictureActivity.java:
```diff
@@ -1,33 +1,81 @@
+
+        if (picturesDirectory.exists()) {
+            File[] listFiles = picturesDirectory.listFiles(new FileFilter() {
+            //Creating list of files from picturesDirectory that will match our File filter conditions
```
Indent

r-oung commented on app/src/main/java/com/example/feelingluckytoday/DisplayPictureActivity.java:
```diff
@@ -1,33 +1,81 @@
+            //Creating list of files from picturesDirectory that will match our File filter conditions
+                @Override
+                public boolean accept(File file) {
+                    if (file.isHidden()) {
+                        return false;
+                    }
+                    if (file.isDirectory()) {
+                        return false;
+                    }
+                    if (file.getPath().endsWith(".jpg")) {
+                        return true;
+                    } else if (file.getPath().endsWith(".png")) {
+                        return true;
+                    } else {
+                        return false;
+                    }
```
You can simplify this:
```java
if (file.isHidden()) {
    return false;
}

if (file.isDirectory()) {
    return false;
}

if (!file.getPath().endsWith(".jpg") && !file.getPath().endsWith(".png")) {
    return false;
}

return true;
```

Keep things consistent, like having all the `returns` in your `if` statements `false` with a single `return true` at the end, make things more clear.

TheW00Xer replied:
I'll keep that in mind. I didn't know about use of && and | | .

Code changed to:
`if (file.isHidden()) { 
     return false; 
} 

if (file.isDirectory()) { 
     return false; 
} 

return file.getPath().endsWith(".jpg") | | file.getPath().endsWith(".png"));`

r-oung commented on app/src/main/java/com/example/feelingluckytoday/DisplayPictureActivity.java:
```diff
@@ -1,33 +1,81 @@
 
-        int[] images = {R.drawable.doge, R.drawable.monke, R.drawable.guine};
-        Random randomNumber = new Random(System.currentTimeMillis());
+        //String[] okFileExtension = {".jpg",".png"};
+        //Declaring value of file name extensions we want the app to use
```
Clean up the code and remove commented code that isn't needed

TheW00Xer replied:
Removed. 
I was keeping the code as reminder for myself to see what I want to try next. 
For example here try to figure out how to use an array of strings as filter value.

r-oung replied:
In that case, you should include an additional comment `// TODO` or `// @TODO`. Android Studio will automatically highlight this and provide you with a filtered list in the sub-menu at the bottom next to `Logcat`. TODO are reminders for developers.

TheW00Xer replied:
Didn't know that. Noted. I'll try it.

r-oung commented on app/src/main/java/com/example/feelingluckytoday/TestButtonActivity.java:
```diff
@@ -0,0 +1,96 @@
+package com.example.feelingluckytoday;
```
This file shouldn't really be included in your commit.

TheW00Xer replied:
Is there a way I can create "Locked" or "Private" commit then where I can include this file and even main activity file with the additional "TEST BUTTON" ?

r-oung replied:
You can use a `.gitignore` file to keep files from being included in your repository, but I think this is different from what you want.

To do what you want, I would suggest creating a `test` folder under `/app/src/main/java/com/example/feelingluckytoday/` and include this `TestButtonActivity.java` there. That makes it clear to other developers that this folder includes only test/mock-up code. Avoid adding test code in your main folders.

r-oung commented on app/src/main/java/com/example/feelingluckytoday/DisplayVideoActivity.java:
```diff
@@ -1,51 +1,123 @@
+        //Declaring MediaController element
+        File videosDirectory = null;
+        //Declaring final directory that will be assigned a path from either main or secondary directory after passing filters
+        File videosDirectory1 = new File(Environment.getExternalStorageDirectory().toString() + "/Lucky/");
```
What is this `/Lucky/` folder for?

TheW00Xer replied:
Lucky folder is just a custom created folder. 
We talked about taking files from local folder and if there is nothing taking files from different one without the app crashing. So I created a folder named Lucky for this purpose. 
So if user doesn't have Lucky folder or there is nothing in it the app will load pictures from default pictures directory.

r-oung reviewed (commented):
Added some comments directly to the code.

In addition:
- Try to simplify the `DisplayVideoActivity.java` in the same way as you've simplified your `DisplayPictureActivity.java` by using file filters.

r-oung commented on app/src/main/java/com/example/feelingluckytoday/DisplayVideoActivity.java:
```diff
@@ -26,79 +24,61 @@
-                        System.out.println("No files with .mp4 extension in 'Lucky' directory");
-                        videosDirectory = videosDirectory2;
-                    }
+                if (!videosDirectory1.isDirectory()) {
```
This is redundant. You already made the same check above.

TheW00Xer replied:
`if (!videosDirectory1.isDirectory())` removed, replaced with `else` in the end.

Lower I'm later checking if Directory1 doesn't exist or doesn't contain usable files to switch from Directory1 to Directory2 if either of those two conditions appears.

r-oung commented on app/src/main/java/com/example/feelingluckytoday/DisplayVideoActivity.java:
```diff
@@ -26,79 +24,61 @@
-                    }
+                if (!videosDirectory1.isDirectory()) {
+                    Toast.makeText(getApplicationContext(), "'Lucky' directory doesn't exist on this device.", Toast.LENGTH_SHORT).show();
+                    System.out.println("'Lucky' directory doesn't exist on this device.");
```
`System.out.println()` is okay to use in Java, but `Log.i()` is more appropriate for Android applications. `Log` gives developers a lot more tools to debug their application, see: https://developer.android.com/studio/debug/am-logcat

TheW00Xer replied:
Noted. 
Changed to `Log.i(TAG,....)`

r-oung commented on app/src/main/java/com/example/feelingluckytoday/DisplayVideoActivity.java:
```diff
@@ -26,79 +24,61 @@
                 }
-            } if (!fileFound) {
+            } if (!videosDirectory1.isDirectory() || listFiles.length <= 0) {
                 if (videosDirectory2.isDirectory()) {
```
Redundant

TheW00Xer replied:
Checking if Directory2 is directory just in case if it isn't I can later on call Toast to tell user that Directory2 also doesn't exist. Even though it should since it's default pictures directory. 

I was just trying to cover any possible scenario.

r-oung commented:
### Pop quiz

A customer comes to you and asks the following:
```
When csv files are outputted through customized Android applications, how can we get those files?

In other words, which folders in the temi are appropriate to be shared?

And could you share procedure steps to create shared folders?
```

If you don't know what a `csv` file is, it's a spreadsheet format.

How would you answer this?

TheW00Xer commented:
> 



> ### Pop quiz
> A customer comes to you and asks the following:
> 
> ```
> When csv files are outputted through customized Android applications, how can we get those files?
> 
> In other words, which folders in the temi are appropriate to be shared?
> 
> And could you share procedure steps to create shared folders?
> ```
> 
> If you don't know what a `csv` file is, it's a spreadsheet format.
> 
> How would you answer this?

An app that creates and outputs files should by default have it's own custom folder where it saves all created files. User shouldn't use any of Temi's folders.
If customer is a developer and want's to access files created by his app, folder created by the app in Android root directory on Temi should be used to hold these files. 
However better approach would be to program the app to use for example "Google Drive" as storage space for such files. So all files created and outputted by this app would be accessible from user's Google Drive storage whenever and wherever needed so user doesn't need to extract them manually from Temi every time if he want's to access those files for example on PC.

r-oung commented:
> However better approach would be to program the app to use for example "Google Drive" as storage space for such files. So all files created and outputted by this app would be accessible from user's Google Drive storage whenever and wherever needed so user doesn't need to extract them manually from Temi every time if he want's to access those files for example on PC.

Can you please provide me with example code of how you would do this?

TheW00Xer commented:
> > However better approach would be to program the app to use for example "Google Drive" as storage space for such files. So all files created and outputted by this app would be accessible from user's Google Drive storage whenever and wherever needed so user doesn't need to extract them manually from Temi every time if he want's to access those files for example on PC.
> 
> Can you please provide me with example code of how you would do this?

`File fileMetadata = new File();
fileMetadata.setName("MyFile.csv");
java.io.File filePath = new java.io.File("files/MyFile.csv");
FileContent mediaContent = new FileContent("text/plain", filePath);
File file = driveService.files().create(fileMetadata, mediaContent)
    .setFields("id")
    .execute();
System.out.println("File ID: " + file.getId());
`
Of course first you need to implement Google Play services in your project and set it up 
https://developers.google.com/identity/sign-in/android/start-integrating

I'm trying to do it within my app somehow but as you know with my skills it will take some time to figure it out. But it is doable and there seem to be quite a few people doing it this way.

r-oung commented:
Sorry if I wasn't clear, but please respond as if I was the client.

> Of course first you need to implement Google Play services in your project and set it up
https://developers.google.com/identity/sign-in/android/start-integrating

Can `Google Play services` be used into temi? Legally?

> I'm trying to do it within my app somehow but as you know with my skills it will take some time to figure it out. But it is doable and there seem to be quite a few people doing it this way.

It's fine to speculate how one might implement something, but when speaking to a customer or our sales people, you should avoid this unless you can immediately provide them with example code. Anything is do-able given enough time and resources.

TheW00Xer commented:
> Sorry if I wasn't clear, but please respond as if I was the client.
> 
> > Of course first you need to implement Google Play services in your project and set it up
> > https://developers.google.com/identity/sign-in/android/start-integrating
> 
> Can `Google Play services` be used into temi? Legally?
> 
> > I'm trying to do it within my app somehow but as you know with my skills it will take some time to figure it out. But it is doable and there seem to be quite a few people doing it this way.
> 
> It's fine to speculate how one might implement something, but when speaking to a customer or our sales people, you should avoid this unless you can immediately provide them with example code. Anything is do-able given enough time and resources.

I see, well then in our particular case if you were the client I would tell you we don't have enough data to provide you with feedback on this issue, since there is no such app in use as of now? (At least none that I know about)
But from the way you formulated the question I deduce that you made a custom app that is producing csv files and you want to access them. If you are able to create this app and install it on Temi you should be able to access Temi's local directories so I would just recommend you to write the app in a way that it creates a custom directory that is easily distinguishable from Temi's directories and use that as a location that you can later access through ADB to extract those files. 
If you asked me if I can give you a solution in form of App that will create csv files and you want to access them later on then my reply would be just the first sentence "We don't have enough data to provide you with feedback on this issue, since there is no such app in use or development as of now."

I always try to look at problems or request as "challenges" because my point of view is exactly like you said "anything is do-able given enough time and resources". I'll probably have to change this approach a bit while I'm at work then since it might create misunderstandings or issues.

As for the Google Play services. From what I found online, Temi is already using "Google speech to text service via their cloud service. https://cloud.google.com/speech-to-text/docs/apis" and that is part of Google Play services. 
Also the tablet itself came pre-installed with Google Play Store just like any regular Android device, it's just not accessible from Temi interface but by creating a shortcut or going to Tablet native screen / settings we can access it so this device is or should be compatible with Google services.
"Google Mobile Services (GMS) is a collection of proprietary applications and application programming interfaces (APIs) services from Google that are typically pre-installed on Android devices, such as smartphones, tablets, and smart TVs. GMS is not a part of the [Android Open Source Project](https://en.wikipedia.org/wiki/Android_Open_Source_Project) (AOSP), which means an Android manufacturer needs to obtain a license from Google in order to legally pre-install GMS on an Android device. This license is provided by Google without any license fees."
(https://en.wikipedia.org/wiki/Google_mobile_services#cite_note-1)

Unless the tablet manufacturer pre-installed these apps and thus used Google services illegally but then the whole product would have illegal software and you shouldn't be using it... 

Also there was an issue/request on Temi's Github from another user about using Google Assistant on Temi (https://github.com/robotemi/sdk/issues/218). Which the Temi team didn't deny using but they encouraged the user to try it and even share more technical details on the integration, they can check it on their side as well. So from this I deduce that it's okay to use Google Play services on/with Temi.

If I'm wrong can you please explain this in greater detail to me, perhaps in private message/mail with some links where I can read more in depth about it?
